contracts/utils/cryptography/draft-ERC7739Utils.sol

```diff
@@ -0,0 +1,288 @@
+     *              keccak256(bytes(name)),
+     *              keccak256(bytes(version)),
+     *              chainId,
+     *              verifyingContract,
+     *              salt,
+     *              keccak256(abi.encodePacked(extensions))
+     *          )
+     *      )
+     *  )
+     *);
+     *
+     * NOTE: The arguments should be those of the verifying application. See {EIP712-_domainSeparatorV4}
+     * and {IERC5267-eip712Domain} for more details of how to obtain these values. Respectively, they
+     * must be obtained from the verifying contract (e.g. an Account) and the application domain.
+     */
+    function toNestedTypedDataHash(
+        bytes32 separator,
+        bytes32 hashedTypedDataNestedStruct
+    ) internal pure returns (bytes32) {
+        return MessageHashUtils.toTypedDataHash(separator, hashedTypedDataNestedStruct);
+    }
+
+    /**
+     * @dev Computes the nested EIP-712 type hash for the given contents type.
+     *
+     * The `contentsTypeName` is the string name in the app's domain before the parentheses
+     * (e.g. Transfer in `Transfer(address to,uint256 amount)`).
+     *
+     * ```solidity
+     * TypedDataSign({contentsTypeName},bytes1 fields,string name,string version,uint256 chainId,address verifyingContract,bytes32 salt,uint256[] extensions){contentsType}
+     * ```
+     *
+     * Requirements:
+     *  - `contentsType` must be a valid EIP-712 type (see {tryValidateContentsType})
+     */
+    // solhint-disable-next-line func-name-mixedcase
+    function NESTED_TYPED_DATA_TYPEHASH(bytes calldata contentsType) internal pure returns (bytes32) {
+        (bool valid, bytes calldata contentsTypeName) = tryValidateContentsType(contentsType);
+        if (!valid) revert InvalidContentsType();
+        return NESTED_TYPED_DATA_TYPEHASH(contentsType, contentsTypeName);
+    }
+
+    /**
+     * @dev Same as {NESTED_TYPED_DATA_TYPEHASH} but with the `contentsTypeName` already validated.
+     *
+     * See {tryValidateContentsType}.
+     */
+    // solhint-disable-next-line func-name-mixedcase
+    function NESTED_TYPED_DATA_TYPEHASH(
+        bytes calldata contentsType,
+        bytes calldata contentsTypeName
+    ) internal pure returns (bytes32) {
+        return
+            keccak256(
+                abi.encodePacked(
+                    "TypedDataSign(",
+                    contentsTypeName,
+                    "bytes1 fields,",
+                    "string name,",
+                    "string version,",
+                    "uint256 chainId,",
+                    "address verifyingContract,",
+                    "bytes32 salt,",
+                    "uint256[] extensions",
+                    ")",
+                    contentsType
+                )
+            );
+    }
+
+    /**
+     * @dev Try to validate the contents type is a valid EIP-712 type.
+     *
+     * A valid `contentsType` is considered invalid if it's empty or it:
+     *
+     * - Starts with a-z or (
```

Comment:
I left [a comment](https://ethereum-magicians.org/t/erc-7739-readable-typed-signatures-for-smart-accounts/20513/4?u=frangio) in the ERC discussion. Disallowing starting lowercase letters is way too restrictive. There’s nothing in EIP-712 that says custom types have to start with an uppercase letter, so we have to assume some applications use lowercase letters. The smart account won’t be usable with those applications which is IMO very serious. I'd recommend against shipping this restriction.

Reply:
One strange thing to me is that it accepts `(` as a first letter. Is that reasonable to not have a name and only have a description ?